authority/provisioner/webhook.go

```diff
@@ -0,0 +1,204 @@
+	"go.step.sm/crypto/sshutil"
+	"go.step.sm/crypto/x509util"
+	"go.step.sm/linkedca"
+)
+
+var ErrWebhookDenied = errors.New("webhook server did not allow request")
+
+type WebhookSetter interface {
+	SetWebhook(string, any)
+}
+
+type WebhookController struct {
+	client       *http.Client
+	webhooks     []*Webhook
+	certType     linkedca.Webhook_CertType
+	TemplateData WebhookSetter
+}
+
+// Enrich fetches data from remote servers and adds returned data to the
+// templateData
+func (wc *WebhookController) Enrich(req *webhook.RequestBody) error {
+	if wc == nil {
+		return nil
+	}
+	for _, wh := range wc.webhooks {
+		if wh.Kind != linkedca.Webhook_ENRICHING.String() {
+			continue
+		}
+		if !wc.isCertTypeOK(wh) {
+			continue
+		}
+		resp, err := wh.Do(wc.client, req, wc.TemplateData)
+		if err != nil {
+			return err
+		}
+		if !resp.Allow {
+			return ErrWebhookDenied
+		}
+		wc.TemplateData.SetWebhook(wh.Name, resp.Data)
+	}
+	return nil
+}
+
+// Authorize checks that all remote servers allow the request
+func (wc *WebhookController) Authorize(req *webhook.RequestBody) error {
+	if wc == nil {
+		return nil
+	}
+	for _, wh := range wc.webhooks {
+		if wh.Kind != linkedca.Webhook_AUTHORIZING.String() {
+			continue
+		}
+		if !wc.isCertTypeOK(wh) {
+			continue
+		}
+		resp, err := wh.Do(wc.client, req, wc.TemplateData)
+		if err != nil {
+			return err
+		}
+		if !resp.Allow {
+			return ErrWebhookDenied
+		}
+	}
+	return nil
+}
+
+func (wc *WebhookController) isCertTypeOK(wh *Webhook) bool {
+	if wc.certType == linkedca.Webhook_ALL {
+		return true
+	}
+	if wh.CertType == linkedca.Webhook_ALL.String() || wh.CertType == "" {
+		return true
+	}
+	return wc.certType.String() == wh.CertType
+}
+
+type Webhook struct {
+	ID                   string `json:"id"`
+	Name                 string `json:"name"`
+	URL                  string `json:"url"`
+	Kind                 string `json:"kind"`
+	DisableTLSClientAuth bool   `json:"disableTLSClientAuth,omitempty"`
+	CertType             string `json:"certType"`
+	Secret               string `json:"-"`
+	BearerToken          string `json:"-"`
+	BasicAuth            struct {
+		Username string
+		Password string
+	} `json:"-"`
+}
+
+func (w *Webhook) Do(client *http.Client, reqBody *webhook.RequestBody, data any) (*webhook.ResponseBody, error) {
+	tmpl, err := template.New("url").Funcs(templates.StepFuncMap()).Parse(w.URL)
+	if err != nil {
+		return nil, err
+	}
+	buf := &bytes.Buffer{}
+	if err := tmpl.Execute(buf, data); err != nil {
+		return nil, err
+	}
+	url := buf.String()
+
+	switch tmpl := data.(type) {
+	case x509util.TemplateData:
+		reqBody.Token = tmpl[x509util.TokenKey]
+	case sshutil.TemplateData:
+		reqBody.Token = tmpl[sshutil.TokenKey]
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+
+	retries := 1
+retry:
+	reqBody.Timestamp = time.Now()
+
+	reqBytes, err := json.Marshal(reqBody)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
I believe the retry should have the same body, so I would move the retry to later.

I'm not sure right now if you will need to recreate the Request or not because the body reader might be consumed if we try to use it twice.

Reply:
58be425